mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,360 @@
+                                 if !inc_event.is_object() {
+                                     continue;
+                                 }
+
+                                 if let Some(id) = inc_event.get("id") {
+                                     let request_id = id.as_u64().unwrap_or_default() as usize;
+
+                                     if let Some(notifier) = response_notifiers.remove(&request_id) {
+                                         let mut res_bytes: Vec<u8> = Vec::new();
+                                         if serde_json::to_writer(&mut res_bytes, &inc_event).is_ok() {
+                                             let response_map = unsafe { &mut *self.responses.0 };
+                                             let _ = response_map.insert(request_id, res_bytes);
+
+                                             notifier.send(()).expect("receiver channel must be alive");
+                                         }
+
+                                     }
+                                 }
+                             }
+                         },
+                         Some(Ok(tokio_tungstenite_wasm::Message::Binary(_))) => continue,
+                         Some(Ok(tokio_tungstenite_wasm::Message::Close(_))) => break,
+                         Some(Err(e)) => {
+                            log::error!("{e}");
+                            return;
+                         },
+                         None => continue,
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn stop_connection_loop(&self) {
+        let mut tx = self.controller_channel.tx.lock().await;
+        tx.send(ControllerMessage::Close)
+            .await
+            .expect("receiver channel must be alive");
+
+        let response_map = unsafe { &mut *self.responses.0 };
+        response_map.clear();
+    }
+
+    pub(crate) fn maybe_spawn_connection_loop(&self, coin: EthCoin) {
+        // if we can acquire the lock here, it means connection loop is not alive
+        if self.connection_guard.try_lock().is_some() {
+            let fut = self.clone().start_connection_loop(None);
+            let settings = AbortSettings::info_on_abort(format!("connection loop stopped for {:?}", self.node.uri));
+            coin.spawner().spawn_with_settings(fut, settings);
+        }
+    }
+
+    pub(crate) fn maybe_spawn_temporary_connection_loop(&self, coin: EthCoin, expires_at: Instant) {
+        // if we can acquire the lock here, it means connection loop is not alive
+        if self.connection_guard.try_lock().is_some() {
+            let fut = self.clone().start_connection_loop(Some(expires_at));
+            let settings = AbortSettings::info_on_abort(format!("connection loop stopped for {:?}", self.node.uri));
+            coin.spawner().spawn_with_settings(fut, settings);
+        }
+    }
```

Comment:
lets reduce code duplication by providing additional private method for `WebsocketTransport`

```rust
    pub(crate) fn maybe_spawn_connection_loop(&self, coin: EthCoin) { self.spawn_connection_loop_helper(coin, None); }

    pub(crate) fn maybe_spawn_temporary_connection_loop(&self, coin: EthCoin, expires_at: Instant) {
        self.spawn_connection_loop_helper(coin, Some(expires_at));
    }

    fn spawn_connection_loop_helper(&self, coin: EthCoin, expires_at: Option<Instant>) {
        // if we can acquire the lock here, it means connection loop is not alive
        if self.connection_guard.try_lock().is_some() {
            let fut = self.clone().start_connection_loop(expires_at);
            let settings = AbortSettings::info_on_abort(format!("connection loop stopped for {:?}", self.node.uri));
            coin.spawner().spawn_with_settings(fut, settings);
        }
    }
```

Reply:
Done